torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp

```diff
@@ -840,45 +798,56 @@ ExprGroup* ExprSegmentationSorter::makeMergedNode(
   return joined_groups;
 }
 
-bool canReduceCA(ExprGroup* group) {
-  IterDomain* g_last_id = nullptr;
-
-  if (group->payload()->ca_domains_.size() > 0) {
-    g_last_id = group->payload()->ca_domains_.back();
-  }
-  if (g_last_id == nullptr) {
+bool canReducePA(ExprGroup* group) {
+  if (group->payload()->pa_domains_.empty()) {
     return false;
   }
 
-  // Compute at can sometimes get in a strange position as the update rules are
-  // not fool proof. All consumers should have a match to this groups inner most
-  // compute at axis, otherwise it should be lowered.
-  for (auto consumer_edge : group->consumerEdges()) {
-    auto consumer = consumer_edge->to;
-    for (auto c_id : consumer->payload()->pa_domains_) {
-      if (GpuLower::current()->caLoopMap().areMapped(c_id, g_last_id)) {
-        return false;
-      }
+  IterDomain* group_pa_last_id = group->payload()->pa_domains_.back();
+
+  // Look through producer edges to see if we can reduce our produce at domain
+  for (auto producer_edge : group->producerEdges()) {
+    auto producer_val = producer_edge->producer_val_;
+    auto consumer_val = producer_edge->consumer_val_;
+
+    // If producer isn't a tensor view it can't be mapped into a producer dim of
+    // this group
+    if (!(consumer_val->isA<TensorView>() && producer_val->isA<TensorView>())) {
+      continue;
     }
-  }
 
-  return true;
-}
+    // If the compute at domains of the producer group is empty, it can't map to
+    // the produce at domains of this group
+    auto producer_group = producer_edge->from;
+    if (producer_group->payload()->ca_domains_.empty()) {
+      continue;
+    }
 
-bool canReducePA(ExprGroup* group) {
-  IterDomain* g_last_id = nullptr;
+    auto producer_tv = producer_val->as<TensorView>();
+    auto consumer_tv = consumer_val->as<TensorView>();
+
+    // If this consumer_tv doesn't map to the last producer domain of this group
+    // it can't decide if it can be reduced
+    bool has_matching_pa = false;
+    for (int i = 0; i < consumer_tv->getMaxProducerPosition(); i++) {
+      if (GpuLower::current()->caLoopMap().areMapped(
+              consumer_tv->axis(i), group_pa_last_id)) {
+        has_matching_pa = true;
+        break;
+      }
+    }
 
-  if (group->payload()->pa_domains_.size() > 0) {
-    g_last_id = group->payload()->pa_domains_.back();
-  }
-  if (g_last_id == nullptr) {
-    return false;
-  }
+    if (!has_matching_pa) {
+      continue;
+    }
 
-  for (auto producer_edge : group->producerEdges()) {
-    auto producer = producer_edge->from;
-    for (auto p_id : producer->payload()->ca_domains_) {
-      if (GpuLower::current()->caLoopMap().areMapped(p_id, g_last_id)) {
+    // If any compute at positions of producers directly map to the last produce
+    // at position it can't be lowered.
+    for (int producer_pos_i = producer_tv->getComputeAtPosition();
```

Comment:
I believe checking with `producer_tv` instead of `ca_domains_` is safer. 

Not sure if I'm understanding correctly, but it seems that after this update the only way we are using `ca_domains_` is taking its size. Is this intended?

Reply:
Yeah, that's a good point. Need to think about if we can just remove ca_domains_ completely.

Reply:
We are using compute at domains validates which producer tensors have the inner most compute at dimension. We would not want a producer merging with a consumer if the edge it's merging on is the inner most loop of that consumer, but not of all consumers. Compute at domain and produce at domain give us a "group view" on the ordering of the loops so we can figure this out. I don't see a way to simplify it.

Reply:
I see. Thanks for giving some thought to simplification. I don't think it should block this PR. So we could leave this part as is and maybe re-visit later.